SuperILC support for scripted test app executions

trylek opened the pull request:
In CoreCLR, each test app has an accompanying cmd script that executes
the app using CoreRun, possibly passing command-line arguments as
required by the particular test case. Previously, SuperIlc used to
blindly execute the native executable - the lack of proper command-line
arguments were causing failures in dozens of tests.

This change tweaks SuperIlc to embrace the execution scripts as a
preferred alternative to directly running the executables [that option
still exists and is exercised in the CoreRT R2R testing]. I have also
simplified bits of the code by directly passing around the BuildOptions
structure rather than propagating multiple arguments representing
various fields on the structure. This also let me better encapsulate
locating the CoreRun executable.

I have replaced CrossgenDirectory with CoreRootDirectory with better
serves our needs as it needs to be passed as the CORE_ROOT environment
variable to the test execution scripts. I have restored Simon's original
code treating --crossgen as a mere flag.

I have modified ParallelRunner.Launch to output to the Console as this
particular message mostly serves as a visual progress indicator when
running the tests, it's no use in the rollup log.

I have slightly trimmed down the ProcessRunner reports: when the process
succeeds, we don't need to report its exit code as it's either implicit
or we can look it up in the scripts. In case of failure, only dump
the hex form of the exit code if it's negative as that's where NT
statuses or HRESULT's get displayed as meaningless big negative integers.

Thanks

Tomas

nattress commented on tests/src/tools/ReadyToRun.SuperIlc/Application.cs:
```diff
@@ -16,25 +16,25 @@ public class Application
 
         private List<string> _mainExecutables;
 
+        private List<string> _executionScripts;
```
When I squint at this file I can almost see factoring out the whole execution script instead of executable launch as a specialization of this class. Ideally the cmd details would be contained in a single super-class to keep complexity of this tool in check. I'm not going to hold up check-in since this is a test tool and not product code.

trylek replied:
Chatted offline. With today functionality making Application itself abstract is complicated because Application can run an arbitrary set of exe and cmd executions. Perhaps the biggest problem is that, after the execution generalizations, the class name "Application" has become a bit of a misnomer as now it rather means "a folder containing given sets of assemblies to compile and execute".

For now I'm proposing renaming the Application class to BuildFolder and the ApplicationSet class to BuildFolderSet, that will make the intent much more obvious. We'll tackle additional abstractions regarding execution of scripts vs. executables in a follow-up change as the exact design for this is not yet completely clear.

trylek commented:
I have renamed Application to BuildFolder as agreed offline with Simon. As an additional small improvement I have split the EXE_FAIL bucket into three sub-buckets - TIMED_OUT, CRASHED and WRONG_EXIT_CODE (where CRASHED is identified based on large negative exit code, for now I randomly picked the constant -1 million as small negative exit codes are commonly used by console apps to report various (typically negative) outcomes. @nattress - could you please take another look?
